gcsio/src/main/java/com/google/cloud/hadoop/gcsio/EventLoggingHttpRequestInitializer.java

```diff
@@ -0,0 +1,89 @@
+import java.time.ZoneOffset;
+import java.time.format.DateTimeFormatter;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Map;
+
+class EventLoggingHttpRequestInitializer implements HttpRequestInitializer {
+  private static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
+  private static final Gson gson = new Gson();
+
+  // Using a ConcurrentHashMap with weak key reference to avoid potential memory leak during failure
+  // scenarios. We do not get an response interceptor callback if there is some failure while
+  // executing the
+  // HttpRequest.
+  private final Map<HttpRequest, HttpRequestResponseTimeTracker> requestTracker =
+      new MapMaker().weakKeys().makeMap();
+
+  @Override
+  public void initialize(HttpRequest request) throws IOException {
+    request.setInterceptor(
+        httpRequest -> requestTracker.put(httpRequest, new HttpRequestResponseTimeTracker()));
+    request.setResponseInterceptor(this::logAndRemoveRequestFromTracking);
+  }
+
+  @VisibleForTesting
+  void logAndRemoveRequestFromTracking(HttpResponse httpResponse) {
+    HttpRequest request = httpResponse.getRequest();
+
+    Map<String, Object> jsonMap = new HashMap<>();
+    jsonMap.put("request_method", request.getRequestMethod());
+    jsonMap.put("request_url", request.getUrl().toString());
+    jsonMap.put("request_headers", request.getHeaders());
+    jsonMap.put("response_headers", httpResponse.getHeaders());
+    jsonMap.put("response_status_code", httpResponse.getStatusCode());
+    jsonMap.put("response_status_message", httpResponse.getStatusMessage());
+    jsonMap.put("thread_name", Thread.currentThread().getName());
+
+    HttpRequestResponseTimeTracker tracker = requestTracker.get(request);
+    if (tracker != null) {
+      requestTracker.remove(request);
+      jsonMap.put("response_time", tracker.getResponseTime());
+      jsonMap.put("request_start_time_utc", tracker.getStartTimeUtc());
+      jsonMap.put("request_finish_time_utc", tracker.getCurrentTimeUtc());
+    } else {
+      jsonMap.put("response_time", Integer.MAX_VALUE);
+      jsonMap.put("unexpected_error", "Zombie request. This is unexpected.");
+    }
+
+    logDetails(jsonMap);
+  }
+
+  @VisibleForTesting
+  void logDetails(Map<String, Object> jsonMap) {
+    logger.atInfo().log(gson.toJson(jsonMap));
+  }
+
+  private static class HttpRequestResponseTimeTracker {
+    private static final DateTimeFormatter dateTimeFormatter =
+        DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSSXXX").withZone(ZoneOffset.UTC);
+
+    private final Stopwatch stopwatch = Stopwatch.createStarted();
+    private final Date startTime = new Date();
+
+    long getResponseTime() {
+      return stopwatch.elapsed(MILLISECONDS);
+    }
+
+    String getStartTimeUtc() {
+      return dateTimeFormatter.format(startTime.toInstant());
+    }
+
+    String getCurrentTimeUtc() {
+      return dateTimeFormatter.format(new Date().toInstant());
+    }
```

Comment:
Do we want to hard code UTC ? why no rely on local time zone, this might conflict for the customers on-prem

Reply:
The main audience for this is GCS team. For them if might be useful to have a standard time, rather than the local time on the machine. Also it is quite easy to convert to any other time zone if required.

Reply:
This doesn't honor the jvm timezone property, and there will be discrepancies. I think we should rely on cloud logging to address the times (event time vs log time). I recommend that we stick to default timezone, instead of hard coding it. This also becomes difficult to evolve with time over multiple releases

Reply:
The timestamps (receiveTimestamp and logTimestamp) from Cloud Logging are there, which is also in UTC. Note that the logTimestamp is the time when the logger actually received the log, which is when the request actually ends. Someone looking at the logs will have to subtract the response_time from the logTimestamp to get when the request was fired. Hence it will be useful (even though it can be derived from latency and log time) to have the start and end times in json_payload itself. Also since the Cloud logging log times are in UTC it is better to have the same in the json_payload